job-server/src/main/scala/spark/jobserver/AkkaClusterSupervisorActor.scala

```diff
@@ -271,25 +273,60 @@ class AkkaClusterSupervisorActor(daoActor: ActorRef, dataManagerActor: ActorRef)
       }
 
     case Terminated(actorRef) =>
-      val name: String = actorRef.path.name
-      logger.info("Actor terminated: {}", name)
-      val contextId = name.split(AkkaClusterSupervisorActor.MANAGER_ACTOR_PREFIX).apply(1)
-      val resp = getDataFromDAO[JobDAOActor.ContextResponse](JobDAOActor.GetContextInfo(contextId))
-      resp match {
-        case Some(JobDAOActor.ContextResponse(Some(c))) =>
-          val state =
-            if (c.state == ContextStatus.Stopping) ContextStatus.Finished else ContextStatus.Killed
-          val contextInfo = ContextInfo(c.id, c.name, c.config, c.actorAddress, c.startTime,
-                Option(DateTime.now()), state, c.error)
-          daoActor ! JobDAOActor.SaveContextInfo(contextInfo)
-          daoActor ! JobDAOActor.CleanContextJobInfos(c.id, DateTime.now())
-        case Some(JobDAOActor.ContextResponse(None)) =>
-          logger.error("No context for deletion is found in the DB")
-        case None =>
-          logger.error("Error occurred after Terminated message was recieved")
-      }
-      cluster.down(actorRef.path.address)
-      jobManagerActorRefs.remove(actorRef.path.toString())
+      handleTerminatedEvent(actorRef)
+  }
+
+  protected def handleTerminatedEvent(actorRef: ActorRef) {
+    val name: String = actorRef.path.name
+    logger.info("Actor terminated: {}", name)
+    val contextId = name.split(AkkaClusterSupervisorActor.MANAGER_ACTOR_PREFIX).apply(1)
+    val resp = getDataFromDAO[JobDAOActor.ContextResponse](JobDAOActor.GetContextInfo(contextId))
+    resp match {
+      case Some(JobDAOActor.ContextResponse(Some(c))) =>
+        val state =
+                if (c.state == ContextStatus.Stopping) ContextStatus.Finished else ContextStatus.Killed
+        (isSuperviseModeEnabled(config, ConfigFactory.parseString(c.config)), state) match {
+            case (true, ContextStatus.Killed) =>
+              setRestartingStateForContextAndJobs(c)
+            case _ =>
+              val contextInfo = c.copy(endTime = Option(DateTime.now()), state = state)
+              daoActor ! JobDAOActor.SaveContextInfo(contextInfo)
+              daoActor ! JobDAOActor.CleanContextJobInfos(c.id, DateTime.now())
+         }
+      case Some(JobDAOActor.ContextResponse(None)) =>
+        logger.error("No context for deletion is found in the DB")
```

Comment:
nit: Do you think that if this error ever happen the reader will need more information? Maybe `contextId` or `name` ?

Reply:
Done